### ISSUE TITLE
ChangeDetectorRef.detach / detectChanges not working as expected

### ISSUE THREAD

**Issue by choeller** [angular/angular#7311]
I want to implement a simple LogViewer, that has a List of Strings as Input and should rerender every 2 seconds. I have several strange effects, when trying to implement this behaviour:

1.) After detaching the  ChangeDetector via ChangeDetectorRef the log-viewer component never refresh again - even after "detectChanges"-call (this is the same setup as here: )

```javascript
@Component({
  selector: 'log-viewer',
  template: `
  <div class="logs">
    <ul>
        <li *ngFor="#log of logs">
            {{log}}
        </li>
    </ul>
</div>
  `,
})
export class LogViewer {
  @Input("logs") logs;
  constructor(private changeDetector: ChangeDetectorRef) {
  }
  ngOnInit() {
    this.changeDetector.detach();
    setInterval(() => {
     // this has no effect at all
      this.changeDetector.detectChanges();
    }, 2000);
  }
  ngOnChanges(change) {
    //this is printed, so there are input binding changes
    console.log("CHANGE", change)
  }
}

```
2.) When detaching the component via the Decorator, the component is rerendered after every change:
```javascript
@Component({
  selector: 'log-viewer',
  template: `...
  `,
  //this has no effect: still renders on each change
  changeDetection: ChangeDetectionStrategy.Detached
})
export class LogViewer {
  @Input("logs") logs;
  constructor(private changeDetector: ChangeDetectorRef) {
  }
  ngOnInit() {
    setInterval(() => {

      this.changeDetector.detectChanges();
    }, 1000);
  }
  ngOnChanges(change) {
    console.log("CHANGE", change)
  }
}
```
This plnkr demonstrates the behaviour: http://plnkr.co/edit/8YQt6PZF5EPkndKgzl82?p=preview

**Comment by choeller** [comment]
So the following "solves" the problem. But is kind of ugly:
```
  ngOnInit() {
    this.changeDetector.detach();
    setInterval(() => {
      this.changeDetector.reattach();
      this.changeDetector.detectChanges();
      this.changeDetector.detach();

    }, 1000);
  }
```

**Comment by zoechi** [comment]
Would be interesting to know how this is supposed to be used. 

AFAIK

    this.changeDetector.detectChanges();

just does change detection and returns the changes found, but doesn't notify Angular about anything.

**Comment by choeller** [comment]
@zoechi According to the documentation this is how it's supposed to work:
https://angular.io/docs/js/latest/api/core/ChangeDetectorRef-class.html

Also when reattaching before the detectChanges it works (only reattaching is not sufficent). So somehow this seems to work, just not when CD is deteched.

**Comment by PascalPrecht** [comment]
@choeller change detector *has* to be reattached to CD is performed for its subtree. The sample in the docs, seems to be buggy. `detectChanges` performs change detection on this tree, however, if your change detector is detached, it won't perform change detection for its subtree.

**Comment by choeller** [comment]
@PascalPrecht Ok, thanks for the reply. So do you see a more elegant solution for triggering redraws manually than doing
```
      this.changeDetector.reattach();
      this.changeDetector.detectChanges();
      this.changeDetector.detach();
```
? I think the example from the docs is a common real world scenario, so it would be cool to have an API for doing this... Also do you have an idea why
```
  changeDetection: ChangeDetectionStrategy.Detached
```
seems to have no effect at all?
Thanks 
Christoph

**Comment by PascalPrecht** [comment]
> So do you see a more elegant solution for triggering redraws manually than doing

I don't know any, but it would just be a facade calling exactly these three functions (unless you want to conditionally check something first)

> Also do you have an idea why

  changeDetection: ChangeDetectionStrategy.Detached
seems to have no effect at all?

That's because `ChangeDetectionStrategy.Detached` is not meant to be used by you at all. The only change detection strategies you can use are `OnPush` and `OnPushObserve`. The latter is experimental and only works in `Dart`. So if you're in TypeScript land `OnPush` (besides `Default`) is the only one you're interested in.

In fact `OnPush` (and `OnPushObserve`) need to be move to a new enum, so they aren't defined in the same as `Checked`, `Detached`, .. etc.

**Comment by ericmartinezr** [comment]
@PascalPrecht OnPushObserve is being dropped (https://github.com/angular/angular/pull/7110), so what's the point on having only one strategy? What's wrong on using any other? Why were those exposed from the beginning? I don't see what's wrong on using them if you know what you're doing.

**Comment by PascalPrecht** [comment]
@ericmartinezr There are two actual "strategies" as in something you can use as a consumer. That's `Default` and `OnPush`.

>  What's wrong on using any other?

Nothing wrong. Simply won't have any effect by design. Other "strategies" are only used internally via apis like `markForCheck()` etc.

>  Why were those exposed from the beginning?

This simply grew historically I'd say.

**Comment by PascalPrecht** [comment]
It's planned to move `Default` and `OnPush` into its own enum so they aren't exposed in the same API.

**Comment by choeller** [comment]
@PascalPrecht Thanks for the insights! Would be cool to have some more comfort when controlling the rendering, but if you know how to do it, it's totally fine as it is... so I will close the ticket for now...

**Comment by gionkunz** [comment]
Well, isn't `Detached` also part of the "public API"? It's certainly what I'm using to do manual change detection in components using the `ChangeDetectorRef`.

**Comment by ridaKourchal** [comment]
Hello,
I have problem with detached is not a function any help
thanks advance

**Comment by DelphiWorlds** [comment]
We're having the same issue as @choeller, but his "solution" kills the performance of our app, using Angular 7.0.3
